client/state/posts/selectors.js

```diff
@@ -149,6 +151,38 @@ export function getSitePostsForQueryIgnoringPage( state, siteId, query ) {
 }
 
 /**
+ * Returns an array of posts for the posts query, including all known queried
+ * pages, preserving hierarchy. Returns null if no posts have been received.
+ * Hierarchy is represented by `parent` and `items` properties on each post.
+ *
+ * @param  {Object} state  Global state tree
+ * @param  {Number} siteId Site ID
+ * @param  {Object} query  Post query object
+ * @return {?Array}        Hierarchical posts for the post query
+ */
+export const getSitePostsHierarchyForQueryIgnoringPage = createSelector(
+	( state, siteId, query ) => {
```

Comment:
>Hmm, I feel like we might be able to offer a default cache key strategy for plain objects, for example JSON.stringify() + md5, and then joined with the other args.

Thinking about this a bit more, to me it makes sense to continue allowing customization of the cache key behavior, specifically for this exact type of use-case (where we already have a serialization strategy for querying).

I also think we should keep the default strategy minimal and performant, and as much as possible, guide people towards simple parameters in their selectors. For these reasons, I'd argue that `Array.prototype.join` is a better choice than `JSON.stringify`.

As an aside, I did some performance testing of the two. `Array.prototype.join` is always faster, but less substantially so for smaller sets of data (which is more realistic for function arguments):

- Large(ish) data set: http://jsperf.com/join-vs-json-stringify
- Small data set: http://jsperf.com/join-vs-json-stringify/2

Thoughts? Keeping in mind that if someone wanted to use `JSON.stringify`, it's as simple as passing:

```jsx
( state, ...args ) => JSON.stringify( args )
```